### DISCUSSION TITLE
Shrink container image size when using Ultralytics with TensorRT

### DISCUSSION THREAD

**Discussion by Xrondev** [ultralytics/ultralytics#23382] [Q&A] [answered]
I built an application inside a container, with ultralytics for the yolo tracking. To have higher throughput, I used the TensorRT for the inference. 

# Problem
The image size explodes after I started to use GPU and TensorRT for the inference, ultralytics installed the TensorRT libs, and seems a CUDA environment is also installed via the pip, here is what I got in my `site-packages`. My final image size came to ~16 GB.
``` shell
site-packages = /app/.venv/lib/python3.13/site-packages
# ...truncated
1.2G	/app/.venv/lib/python3.13/site-packages/cv2
1.7G	/app/.venv/lib/python3.13/site-packages/torch
4.3G	/app/.venv/lib/python3.13/site-packages/nvidia
5.7G	/app/.venv/lib/python3.13/site-packages/tensorrt_libs
```

I searched for few sources and asked GPT for a explanation, seems a cuda library has been installed in the /app/.venv/lib/python3.13/site-packages/nvidia; However, my base image already provided the cuda library and environment in it. 

This base image is being used, and runtime of the following image is used to build the final image
``` Dockerfile
FROM nvidia/cuda:12.8.0-cudnn-devel-ubuntu22.04 
```

**Is there any suggestion to shrink the image size?** Moving to a raw TensorRT with bytetrack and no longer use ultralytics can be a solution, but I am still seeking for a solution with ultralytics, which is more convenient to use the yolo and the tracker. 

I am not quite sure if the docker image of ultralytics provided proper global cuda libs, with nvidia codec abilities available (I primarily doing video works and GPU decode is preferred)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Xrondev, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it (for example: a minimal `Dockerfile`, install commands, and the exact steps that cause the large `site-packages` footprint).

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖🙂—an Ultralytics engineer will also assist soon. To help us understand the container size jump with TensorRT/GPU installs, please share:
-Your minimal `Dockerfile` (including base image, Python version, and all `pip install` lines)
-Your exact `ultralytics` version and how TensorRT is being enabled/installed
-The output of `pip freeze` (or at least the `torch*`, `nvidia*`, and `tensorrt*` lines)
-The commands you used to measure the directory sizes (so we can reproduce the same breakdown)

**Reply by glenn-jocher** [reply, marked as answer]
What you’re seeing is expected: the large `site-packages/nvidia` and `tensorrt_libs` come from `pip`-installed PyTorch/TensorRT wheels bundling CUDA/TensorRT runtime deps (even if your base image already has CUDA), as NVIDIA notes in their [TensorRT installation docs](https://docs.nvidia.com/deeplearning/tensorrt/10.13.0/installing-tensorrt/installing.html); the most reliable way to shrink is to avoid those wheels (use system/NGC-provided TensorRT+PyTorch, then `pip install ultralytics --no-deps` only if you’ve already satisfied deps) and use a multi-stage build so your final stage is a `*-runtime` CUDA image plus just your app and exported `*.engine` (Ultralytics will load it normally per the [TensorRT integration guide](https://docs.ultralytics.com/integrations/tensorrt/)). For GPU video decode, that’s typically down to your FFmpeg/OpenCV build flags rather than Ultralytics itself, so you’ll want to validate that in your base image (our [Docker quickstart guide](https://docs.ultralytics.com/guides/docker-quickstart/) covers the recommended container patterns).
